Update Roslyn snapshot when editorconfig is updated. for now, it only…

heejaechang opened the pull request:
… updates when editorconfig is saved.

### Customer scenario

User updates editorconfig and save the file but the change doesn't get picked up by other files opened in VS until files are closed and reopened.

### Bugs this fixes

https://github.com/dotnet/roslyn/issues/15003

### Workarounds, if any

modify the file or close and reopen the file.

### Risk

Low.

### Performance impact

Low

### Is this a regression from a previous update?

No

### Root cause analysis

We consider options as a part of snapshot. but we didn't update snapshots when editorconfig options are changed. for now, we support it for opened files, but for full experience, it require editorconfig work from compiler.

### How was the bug found?

Dogfooding, Feedbacks

jasonmalinowski commented:
@heejaechang Will look in detail tomorrow, but gist seems OK.

jasonmalinowski commented:
(wrong button)

jasonmalinowski commented on src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs:
```diff
@@ -1051,5 +1051,17 @@ public OptionSet Options
+        internal Solution WithOptionChanged()
+        {
+            // just create new snapshot
+            return new Solution(_state);
```
@CyrusNajmabadi Basically, Document holds onto a snapshot of the file's options, so that way that's immutable. When the options change, we aren't (currently) creating a new snapshot so values would be stale. This is at least working around that problem before we integrate .editorconfig more cleanly into workspaces, which is currently underway.

CyrusNajmabadi replied:
@jasonmalinowski What i meant was: we need to document that on the call. :)

jinujoseph commented:
approved to merge via [link](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/602851) for 15.7.Preview5

jinujoseph commented:
one more review pending

sharwell commented on src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs:
```diff
@@ -47,7 +50,13 @@ private void Workspace_DocumentClosed(object sender, DocumentEventArgs e)
+                        {
+                            var context = t.Result;
+
+                            context.CodingConventionsChangedAsync -= OnCodingConventionsChangedAsync;
```
❓ Does this matter if we are disposing of the object anyway?

olegtk replied:
This is an editor's component, but please be a good citizen still :)
I wonder if you need to unsubscribe before scheduling a task though, what if the .editorconfig file has changed in between lines 52 and 57 and OnCodingConventionsChangedAsync gets called?

sharwell commented on src/EditorFeatures/Core.Wpf/Options/EditorConfigDocumentOptionsProvider.cs:
```diff
@@ -119,8 +128,24 @@ private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
                 defaultValue: EmptyCodingConventionContext.Instance);
         }
+
+        private Task OnCodingConventionsChangedAsync(object sender, CodingConventionsChangedEventArgs arg)
```
❓ @olegtk Is this event ever invoked more often than one invocation for one change of .editorconfig on disk?

heejaechang replied:
if you are asking whether changing 1 editconfig file that affects multiple files will cause multiple events to be fired. then I tested it, and they do fire events for each file that got affected by the edit config file.

olegtk replied:
there should be 1 event per file change (as raised by the VS file watcher service), and it's fired on each affected context instance.

heejaechang commented:
alright. thank you every one. addressed some PR feedbacks.

tagging @jinujoseph 

first, this PR only update snapshot if open file is affected by editor config file change. closed file, it is not easy to do. that will be automatically supported once we have full editor config support from compiler.

second, I did event aggregation. this is good for perf but now we have chance that if users use options within 50ms since last editconfig change events, they will not see the option changes.

heejaechang commented:
opened tracking issue 
https://github.com/dotnet/roslyn/issues/26250

heejaechang commented:
@dotnet/roslyn-infrastructure do I need to keep rerun integration test until it pass or is it okay if one of test (release) passed?

heejaechang commented:
retest windows_debug_vs-integration_prtest please
